src/subprocess-posix.cc

```diff
@@ -119,10 +120,32 @@ bool Subprocess::Start(SubprocessSet* set, const string& command) {
     Fatal("posix_spawnattr_setflags: %s", strerror(err));
 
   const char* spawned_args[] = { "/bin/sh", "-c", command.c_str(), NULL };
+  if (!use_console_) {
+    if (pipe(input_pipe) < 0)
+      Fatal("pipe: %s", strerror(errno));
+    SetCloseOnExec(input_pipe[0]);
+    err = posix_spawn_file_actions_adddup2(&action, input_pipe[0], 0);
+    if (err != 0)
+      Fatal("posix_spawn_file_actions_adddup2: %s", strerror(err));
+    err = posix_spawn_file_actions_addclose(&action, input_pipe[0]);
+    if (err != 0)
+      Fatal("posix_spawn_file_actions_addclose: %s", strerror(err));
+    err = posix_spawn_file_actions_addclose(&action, input_pipe[1]);
+    if (err != 0)
+      Fatal("posix_spawn_file_actions_addclose: %s", strerror(err));
+    spawned_args[1] = NULL;
+  }
   err = posix_spawn(&pid_, "/bin/sh", &action, &attr,
         const_cast<char**>(spawned_args), environ);
   if (err != 0)
     Fatal("posix_spawn: %s", strerror(err));
+  
+  if (!use_console_) {
+    ssize_t written = write(input_pipe[1], command.c_str(), command.size());
```

Comment:
Does that work for large commands?

Most systems use a default pipe buffer size of 64 kiB, some even smaller (see https://github.com/afborchert/pipebuf) so this write() call will block infinitely if `command.size()` is larger than that (and the bug referenced by this PR talks about command sizes of 128 kiB).

Linux has `fcntl(F_SETPIPE_SZ)` to change that the pipe buffer size, but I am not aware of similar features on other systems. Even the Linux pipe manpage tsates that applications should not rely on a specific capacity.

It might be more portable and simple to write long commands to a file that is passed as the first argument to the shell instead. And that would also work for console processes too.

Reply:
friendly ping?

Reply:
Would a temp file-based approach require us to parse the command string just to check the argument limit, to avoid generating a command file for every single task? The alternative pessimistic scan for space separators seems simpler but could lead to various edge cases.